<center><font size=10>Medical Diagnosis Assistant </center></font>
<center><font size=6>Using </center></font>
<center><font size=6> RAG and Large Language Models</center></font>

<center><p float="center">
  <img src="https://blog.pepid.com/wp-content/uploads/2023/03/MicrosoftTeams-image-8-980x549.jpg" width="720"/>
</p></center>

## Project Introduction

The Medical Diagnosis Chatbot is an Artificial Intelligence (AI) application developed to answer medical questions by using Large Language Models (LLMs) and Retrieval-Augmented Generation (RAG). The system uses a medical reference document as its knowledge source and generates responses based on the information available in that document. The medical knowledge base is derived from the Merck Manual of Diagnosis & Therapy, which contains detailed information on diseases, symptoms, diagnosis, treatments, and clinical guidelines.

Unlike a traditional chatbot that relies only on pre-trained knowledge, this chatbot retrieves relevant medical information from the uploaded document before generating an answer. This improves the accuracy and reliability of responses.

## Project Objective

The primary objective of this project is to build an AI-powered Medical Diagnosis Chatbot capable of answering users' medical questions accurately. The project evaluates two approaches:



*  LLM Approach: The Llama-2 model answers questions directly using prompt engineering.
*  RAG Approach: The chatbot retrieves relevant information from the medical PDF and uses it as context for the Llama-2 model to generate more accurate responses.

The objective is to demonstrate that the RAG approach produces more reliable, document-grounded answers than using the LLM alone.

##Problem Statement

Large Language Models can generate fluent responses, but they may produce incorrect or fabricated information (hallucinations), especially in sensitive domains such as healthcare. Medical information must be accurate and evidence-based. To address this challenge, the project integrates a Retrieval-Augmented Generation (RAG) pipeline that retrieves relevant content from the medical reference document before response generation. This ensures that answers are based on the uploaded medical knowledge rather than only the model's pre-trained knowledge.

##Data discription

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

**Common Questions to Answer**

1. **Critical Care Protocols:** "What is the protocol for managing sepsis in a critical care unit?"

2. **General Surgery:** "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"

3. **Dermatology:** "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"

4. **Neurology:** "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"


##Overall Project Workflow


```
                     Medical PDF
                         │
                         ▼
                  Load Document
                         │
                         ▼
                 Text Preprocessing
                         │
                         ▼
                  Split into Chunks
                         │
                         ▼
             Generate Vector Embeddings
                         │
                         ▼
              Store in FAISS Database
                         │
         ┌───────────────┴───────────────┐
         │                               │
         ▼                               ▼
     LLM Approach                  RAG Approach
         │                               │
         ▼                               ▼
 Prompt Engineering          Retrieve Relevant Chunks
         │                               │
         └───────────────┬───────────────┘
                         ▼
                    Llama-2 Model
                         │
                         ▼
                  Generated Response
```


##System Architecture

```
                +-----------------------+
                |   Medical PDF Dataset |
                +-----------+-----------+
                            |
                            v
                  +-------------------+
                  |  PDF Loader       |
                  +-------------------+
                            |
                            v
                  +-------------------+
                  | Text Splitter     |
                  +-------------------+
                            |
                            v
                  +-------------------+
                  | Embedding Model   |
                  +-------------------+
                            |
                            v
                  +-------------------+
                  | FAISS Database    |
                  +-------------------+
                            |
User Question --------------+
                            |
                            v
                  +-------------------+
                  | Retriever         |
                  +-------------------+
                            |
                            v
                  +-------------------+
                  | Llama-2 (LLM)     |
                  +-------------------+
                            |
                            v
                  Final Medical Answer
```

## Installing and Importing Necessary Libraries and Dependencies

In [ ]:
# ==========================================
# Install Required Libraries
# ==========================================

!pip install -q \
langchain==0.3.27 \
langchain-community==0.3.27 \
chromadb==1.0.15 \
pymupdf==1.26.3 \
tiktoken==0.9.0 \
datasets==4.0.0 \
evaluate==0.4.5 \
langchain-openai==0.3.30 \
langchain-huggingface \
sentence-transformers \
huggingface-hub \
transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 20.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 83.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.5/19.5 MB 93.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 79.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 72.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 7.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 74.4/74.4 kB 7.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 27.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 MB 123.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 461.3/461.3 kB 39.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.2/19.2 MB 88.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 948.6/948.6 

In [ ]:
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 \
pip install -q llama-cpp-python==0.2.45 \
--force-reinstall --upgrade --no-cache-dir

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.7/36.7 MB 155.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 287.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.9/134.9 kB 365.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.7/16.7 MB 238.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.6/45.6 kB 314.3 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langgraph-sdk 0.4.2 requires langchain-core<2,>=1.4.0, but you have langchain-core 0.3.86 which is incompatible.
google-adk 2.4.0 requires opentelemetry-api<=1.42.1,>=1.39, but you have opentelemetry-api 1.44.0 which is incompatible.
google-adk 

*   llama-cpp-python
*   huggingface_hub
*   langchain
*   faiss-cpu
*   sentence-transformers
*   transformers

These libraries prepare the development environment for both the LLM and RAG approaches.

In [ ]:
# ==========================================
# Import Required Libraries
# ==========================================

import os

from google.colab import userdata

# PDF Loader
from langchain_community.document_loaders import PyPDFLoader

# Text Splitter (NEW IMPORT)
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Embeddings
from langchain_huggingface import HuggingFaceEmbeddings

# Vector Store
from langchain_community.vectorstores import FAISS

In [ ]:
from langchain_community.embeddings import HuggingFaceEmbeddings
from langchain_community.vectorstores import Chroma
from sentence_transformers import SentenceTransformer

## **LLM with Prompt Engineering Response**

#### **Download LLaMA-2 13B Chat Model**

In [ ]:
from huggingface_hub import hf_hub_download

model_name_or_path = "TheBloke/Llama-2-13B-chat-GGUF"
model_basename = "llama-2-13b-chat.Q5_K_M.gguf"

model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
)

llama-2-13b-chat.Q5_K_M.gguf: reconstructing file:   0%|          |  0.00B / 9.23GB            

llama-2-13b-chat.Q5_K_M.gguf: downloading bytes:           |  0.00B            

The Llama-2-13B Chat GGUF model was downloaded once and stored locally, allowing it to be used without requiring an OpenAI API key or internet connection during inference.

This downloaded model serves as the language generation engine of the chatbot.

#### **Initialize LLaMA Model with Configuration**

In [ ]:
from llama_cpp import Llama

lcpp_llm = Llama(
    model_path=model_path, # Path to the downloaded GGUF model
    n_threads=4,           # Number of CPU threads to use
    n_batch=512,           # Batch size for prompt processing
    n_gpu_layers=-1,       # Number of layers to offload to GPU (-1 for all)
    n_ctx=4096             # Context window
)

llama_model_loader: loaded meta data with 19 key-value pairs and 363 tensors from /root/.cache/huggingface/hub/models--TheBloke--Llama-2-13B-chat-GGUF/snapshots/4458acc949de0a9914c3eab623904d4fe999050a/llama-2-13b-chat.Q5_K_M.gguf (version GGUF V2)
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = LLaMA v2
llama_model_loader: - kv   2:                       llama.context_length u32              = 4096
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 5120
llama_model_loader: - kv   4:                          llama.block_count u32              = 40
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 13824
llama_model_loader: - kv   6:                 llama.rope.dimension_

After downloading the model, it is loaded into memory using the llama-cpp-python library. During loading, parameters such as CPU threads, GPU layers, batch size, and context window are configured to optimize inference performance.

Once initialized, the model is ready to process prompts and generate responses.

##Prompt Engineering

In [ ]:
# Provides an example and answer anchor to guide the model in giving concise, evidence-based responses without echoing the question.
system_prompt = """
You are a medical expert AI assistant. Respond with concise, evidence-based answers.
"""

user_prompt = """
Example:
Q: What are the common symptoms of appendicitis?
A: Common symptoms include abdominal pain (usually starting near the navel), nausea, vomiting, and fever.
References: Mayo Clinic, UpToDate

Now answer:
Q: What is the protocol for managing sepsis in a critical care unit?
A:
"""


Prompt Engineering is the process of designing instructions that guide the LLM to generate high-quality responses.

The prompt consists of:

*   System Prompt
*   User Question
*   Answer Template

The system prompt instructs the model to behave like a medical assistant and provide concise, evidence-based answers.

#### **Response Function**

In [ ]:
# function to generate, process, and return the response from the LLM
def prompt_engineering_response(user_prompt):
    # Put the system message first, then the user question, then anchor with "Answer:"
    prompt = f"""{system_prompt}

Question: {user_prompt}

Answer:"""

    # Generate a response from the LLaMA model
    response = lcpp_llm(
        prompt=prompt,
        max_tokens=500, # Max number of tokens to generate
        temperature=0.3, # Sampling temperature
        top_p=0.95, # Top-p sampling
        stop=["Q:", "\n"], # Stop generating when "Q:" or a new line is encountered
        echo=False # Do not echo the prompt in the output
    )

    # Extract and return the response text
    response_text = response["choices"][0]["text"].strip()
    return response_text


When a user enters a medical question, the prompt is sent to the Llama-2 model. The model analyzes the prompt and generates a response based on its learned medical knowledge.

Generation parameters such as:

* Temperature - 0.3
* Top-p Sampling - 0.95
* Maximum Tokens - 500

are used to control the quality and diversity of the output.

###LLM Workflow

```
User Question
      │
      ▼
Prompt Engineering
      │
      ▼
Llama-2 Model
      │
      ▼
Generate Response
      │
      ▼
Display Output
```

## Question Answering using LLM with Prompt Engineering

### Question 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
question_1 = "What is the protocol for managing sepsis in a critical care unit?"

### Question 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
question_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"

### Question 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
question_3= "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"

### Question 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
question_4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"

###LLM Architecture

```
            User
              │
              ▼
      Medical Question
              │
              ▼
     Prompt Engineering
              │
              ▼
       Llama-2 13B Model
              │
              ▼
      Generated Response
              │
              ▼
         Display Result
```

#### **Create and Display Results DataFrame**

In [ ]:
import pandas as pd

In [ ]:
prompt_result_df = pd.DataFrame({
    "questions": [question_1, question_2, question_3, question_4],
    "prompt_Engineering_responses": [
         prompt_engineering_response(question_1),
         prompt_engineering_response(question_2),
         prompt_engineering_response(question_3),
         prompt_engineering_response(question_4)
    ] })

# Display the DataFrame
prompt_result_df.head()


llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =      28.98 ms /    42 runs   (    0.69 ms per token,  1449.48 tokens per second)
llama_print_timings: prompt eval time =     572.20 ms /    48 tokens (   11.92 ms per token,    83.89 tokens per second)
llama_print_timings:        eval time =    2291.78 ms /    41 runs   (   55.90 ms per token,    17.89 tokens per second)
llama_print_timings:       total time =    3063.28 ms /    89 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =      31.50 ms /    44 runs   (    0.72 ms per token,  1396.96 tokens per second)
llama_print_timings: prompt eval time =     344.56 ms /    37 tokens (    9.31 ms per token,   107.38 tokens per second)
llama_print_timings:        eval time =    2388.88 ms /    43 runs   (   55.56 ms per token,    18.00 tokens per second)
llama_print_timings:       total time =    2949.54 ms /    80 

,questions,prompt_Engineering_responses
0,What is the protocol for managing sepsis in a ...,The Surviving Sepsis Campaign (SSC) guidelines...
1,"What are the common symptoms for appendicitis,...","Appendicitis is inflammation of the appendix, ..."
2,What are the effective treatments or solutions...,"Sudden patchy hair loss, also known as alopeci..."
3,What treatments are recommended for a person w...,The treatment approach for a person with a phy...


The generated responses are stored in a Pandas DataFrame, where each question is paired with its corresponding answer. This tabular format makes it easier to compare responses and evaluate the performance of the LLM approach.

###Data Flow

```
User Question
      │
      ▼
Create Prompt
      │
      ▼
Llama-2 Inference
      │
      ▼
Generate Answer
      │
      ▼
Store in DataFrame
```

####Observation

The LLM approach successfully generates natural-language responses using only the Llama-2 model. While the responses are fluent and informative, they are not grounded in the uploaded medical document, making them potentially less reliable than the RAG approach for document-specific question answering.

## **RAG Response**

The Retrieval-Augmented Generation (RAG) approach combines information retrieval with a Large Language Model (LLM) to generate accurate and context-aware responses. Unlike the LLM approach, which answers questions using only its pre-trained knowledge, the RAG approach first retrieves relevant information from the uploaded Merck Manual medical PDF and then uses that information as context for response generation. This makes the chatbot more reliable and reduces the chances of hallucinations.

## **Data Preparation for RAG**

### **Loading the data**

In [ ]:
# Mount Google Drive to the /content/drive directory to access the files
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


The first step in the RAG pipeline is loading the medical PDF using PyPDFLoader. The loader extracts the text from each page of the document and converts it into LangChain Document objects while preserving metadata such as the page number and source file. This prepares the medical document for further processing.

In [ ]:
from langchain.document_loaders import PyMuPDFLoader

# Load Merck Manual PDF
pdf_path = "/content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf"
loader = PyMuPDFLoader(pdf_path)

# Load as LangChain Documents
document = loader.load()

print(f"Total pages loaded: {len(document)}")



Total pages loaded: 4114


Loading the Merk manual document contains the Medical diagnasis manual of 4114 pages

####Work Flow

```
Medical PDF
      │
      ▼
PyPDFLoader
      │
      ▼
Document Objects
```

### Data Overview

Display the content of page number 16 and 17

In [ ]:
for i in range(15,17):
    print(f"Page {i+1}:")
    print(document[i].page_content)

Page 16:
degree. The book received critical acclaim and sold over 2 million copies. The Second Home Edition was
released in 2003. Merck's commitment to providing comprehensive, understandable medical information
to all people continued with The Merck Manual Home Health Handbook, published in 2009.
The Merck Manual of Health & Aging , published in 2004, continued Merck's commitment to education
and geriatric care, providing information on aging and the care of older people in words understandable
by the lay public.
In 2008, The Merck Manual of Patient Symptoms  was introduced to complement The Merck Manual
and was intended to help newcomers to clinical diagnosis approach patients who present with certain
common symptoms.
As part of its commitment to ensuring that all who need and want medical information can get it, Merck
provides the content of these Merck Manuals on the web for free (www.merckmanuals.com). Registration
is not required, and use is unlimited. The web publications are co

## **Data Chunking**

Split the document into Chunks and display the total chunks

In [ ]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

# Initialize the text splitter
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=520,  # You can adjust this value based on your needs
    chunk_overlap=50, # You can adjust this value based on your needs
    length_function=len,
    is_separator_regex=False,
)

# Split the document into chunks
docs = text_splitter.split_documents(document)

print(f"Total chunks: {len(docs)}")



Total chunks: 30717


The extracted medical text is divided into smaller overlapping chunks using RecursiveCharacterTextSplitter. Chunking makes it easier to generate embeddings and retrieve relevant information efficiently. The overlap between chunks ensures that important context is not lost when splitting the document.

*   Chunk size = 520
*   Chunk overlap = 50



####Work Flow

```
Medical Document
        │
        ▼
RecursiveCharacterTextSplitter
        │
        ▼
Small Text Chunks
```

In [ ]:
chunks = text_splitter.split_documents(document)

### Embedding

Generate Vector Embeddings for Text Chunks Using OpenAI

In [ ]:
embedding = HuggingFaceEmbeddings(
    model_name="sentence-transformers/all-MiniLM-L6-v2"
)

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

Each text chunk is converted into a numerical vector called an embedding using the Hugging Face embedding model sentence-transformers/all-MiniLM-L6-v2. These embeddings capture the semantic meaning of the text, allowing the system to compare the similarity between the user's question and the document content.

In [ ]:
chunk = chunks[0].page_content

embedding = embedding.embed_query(chunk)

print("Embedding Length:", len(embedding))

Embedding Length: 384


####Work Flow

```
Text Chunks
      │
      ▼
Embedding Model
      │
      ▼
Vector Embeddings
```

In [ ]:
from langchain.vectorstores import Chroma
# Re-initialize the embedding object as it was overwritten in a previous cell
#persist_directory="/content/drive/MyDrive/project 2"
# Building the vector store and saving it to disk for future use
vectorstore = Chroma.from_documents(
    documents=docs,
    embedding=embedding,
    persist_directory="."
)

The generated embeddings are stored in a FAISS vector database. FAISS enables fast similarity searches by indexing the embeddings. Instead of searching through the entire PDF, the chatbot searches the vector database to quickly identify the most relevant document chunks.

In [ ]:
print(type(vectorstore))

<class 'langchain_community.vectorstores.chroma.Chroma'>


In [ ]:
query = "What is diabetes?"

results = vectorstore.similarity_search(
    query,
    k=3
)

In [ ]:
for i, doc in enumerate(results, start=1):
    print("=" * 80)
    print(f"Result {i}")
    print("=" * 80)
    print(doc.page_content)
    print()

Result 1
characterized by low plasma glucose level, symptomatic sympathetic nervous system
stimulation, and CNS dysfunction. Many drugs and disorders cause it. Diagnosis requires blood
tests done at the time of symptoms or during a 72-h fast. Treatment is provision of glucose
The Merck Manual of Diagnosis & Therapy, 19th Edition
Chapter 99. Diabetes Mellitus & Disorders of Carbohydrate Metabolism
1021
rvssenthil@gmail.com
7RPZD4O59E
This file is meant for personal use by rvssenthil@gmail.com only.

Result 2
characterized by low plasma glucose level, symptomatic sympathetic nervous system
stimulation, and CNS dysfunction. Many drugs and disorders cause it. Diagnosis requires blood
tests done at the time of symptoms or during a 72-h fast. Treatment is provision of glucose
The Merck Manual of Diagnosis & Therapy, 19th Edition
Chapter 99. Diabetes Mellitus & Disorders of Carbohydrate Metabolism
1021
rvssenthil@gmail.com
7RPZD4O59E
This file is meant for personal use by rvssenthil@gmail.com

In [ ]:
for i, doc in enumerate(results, start=1):
    print(f"Result {i}")
    print("Page:", doc.metadata["page"])
    print("Source:", doc.metadata["source"])
    print()

Result 1
Page: 1030
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf

Result 2
Page: 1030
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf

Result 3
Page: 1010
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf



In [ ]:
query = "What causes hypertension?"

results = vectorstore.similarity_search(
    query,
    k=3
)

for doc in results:
    print(doc.page_content[:500])
    print("\n" + "-" * 80 + "\n")

renovascular disease (see p. 2077), pheochromocytoma, Cushing's syndrome, primary aldosteronism,
congenital adrenal hyperplasia, hyperthyroidism, myxedema, and coarctation of the aorta. Excessive
alcohol intake and use of oral contraceptives are common causes of curable hypertension. Use of
sympathomimetics, NSAIDs, corticosteroids, cocaine, or licorice commonly contributes to hypertension.
Pathophysiology
Because BP equals cardiac output (CO) × total peripheral vascular resistance (TPR), pathog

--------------------------------------------------------------------------------

Chapter 208. Arterial Hypertension
Introduction
Hypertension is sustained elevation of resting systolic BP (≥ 140 mm Hg), diastolic BP (≥ 90 mm
Hg), or both. Hypertension with no known cause (primary; formerly, essential hypertension) is
most common. Hypertension with an identified cause (secondary hypertension) is usually due
to a renal disorder. Usually, no symptoms develop unless hypertension is severe or long

####Work Flow

```
Embeddings
      │
      ▼
FAISS Index
      │
      ▼
Vector Database
```

### Retriever

Retrieval and Response Generation using Vector Search

In [ ]:
# Create the retriever
retriever = vectorstore.as_retriever(
    search_type="similarity",
    search_kwargs={"k": 5}
)

The Retriever receives the user's question, converts it into an embedding, and performs a similarity search in the FAISS database. It retrieves the most relevant document chunks that closely match the user's query. These retrieved chunks serve as the context for the LLM.

In [ ]:
print(type(retriever))

<class 'langchain_core.vectorstores.base.VectorStoreRetriever'>


In [ ]:
question = "What is diabetes?"

docs = retriever.invoke(question)

In [ ]:
for i, doc in enumerate(docs, start=1):
    print("=" * 80)
    print(f"Document {i}")
    print("=" * 80)

    print(doc.page_content[:700])
    print()

Document 1
characterized by low plasma glucose level, symptomatic sympathetic nervous system
stimulation, and CNS dysfunction. Many drugs and disorders cause it. Diagnosis requires blood
tests done at the time of symptoms or during a 72-h fast. Treatment is provision of glucose
The Merck Manual of Diagnosis & Therapy, 19th Edition
Chapter 99. Diabetes Mellitus & Disorders of Carbohydrate Metabolism
1021
rvssenthil@gmail.com
7RPZD4O59E
This file is meant for personal use by rvssenthil@gmail.com only.

Document 2
characterized by low plasma glucose level, symptomatic sympathetic nervous system
stimulation, and CNS dysfunction. Many drugs and disorders cause it. Diagnosis requires blood
tests done at the time of symptoms or during a 72-h fast. Treatment is provision of glucose
The Merck Manual of Diagnosis & Therapy, 19th Edition
Chapter 99. Diabetes Mellitus & Disorders of Carbohydrate Metabolism
1021
rvssenthil@gmail.com
7RPZD4O59E
This file is meant for personal use by rvssenthil@gmail

In [ ]:
for i, doc in enumerate(docs, start=1):
    print(f"Document {i}")
    print("Page:", doc.metadata.get("page"))
    print("Source:", doc.metadata.get("source"))
    print()

Document 1
Page: 1030
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf

Document 2
Page: 1030
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf

Document 3
Page: 1010
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf

Document 4
Page: 1010
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf

Document 5
Page: 2554
Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf



####Work Flow

```
User Question
      │
      ▼
Question Embedding
      │
      ▼
FAISS Search
      │
      ▼
Top Relevant Chunks
```

In [ ]:
medical_system_message = """
You are an AI assistant designed to support healthcare professionals by providing evidence-based, concise, and accurate responses using authoritative medical sources, such as the Merck Manuals.

Your goal is to help clinicians, researchers, and healthcare teams quickly access reliable medical knowledge to improve patient outcomes, support decision-making, and reduce information overload.

User input will include context extracted from trusted medical sources. This context will begin with the token:

###Context
The context may include excerpts from the Merck Manuals, clinical guidelines, or peer-reviewed medical literature, including titles, sections, authors, and other relevant metadata.

When crafting your response:
- Use only the provided context to answer the question.
- Provide concise, clinically relevant, and accurate answers.
- Include the source (title, section, and page/section reference) when applicable.
- If the context does not contain relevant information, respond: "Sorry, this is out of my knowledge base."
- Do NOT provide personal medical advice or treatment recommendations outside of the context.
- Maintain a professional, neutral, and safe tone appropriate for healthcare communication.

Example response format:

Answer:
[Answer based on context]

Source:
[Source title, section, page]
"""


In [ ]:
medical_user_message_template = """
###Context
Here are relevant excerpts from the Merck Manuals or other authoritative medical sources:
{context}

###Question
{question}
"""


After retrieval, the relevant document chunks are combined with the user's question to form a structured prompt. This prompt instructs the Llama-2 model to answer the question using only the retrieved context, ensuring that the response is grounded in the uploaded medical document.

####Work Flow

```
Retrieved Context
        +
User Question
        │
        ▼
Prompt Engineering
        │
        ▼
LLM Prompt
```

### Response Function

In [ ]:
# Function to retrieve relevant context and generate a RAG response
def generate_rag_response(user_input, retriever,
                          system_message, user_message_template,
                          k=5, max_tokens=500,
                          temperature=0.3, top_p=0.95):

    # Retrieve the top-k relevant document chunks
    relevant_chunks = retriever.invoke(user_input)

    if not relevant_chunks:
        return "Sorry, this is out of my knowledge base."

    # Combine retrieved chunks with source information
    context_for_query = "\n\n".join(
        [
            f"Source: {doc.metadata.get('source', 'Unknown')}\n{doc.page_content}"
            for doc in relevant_chunks
        ]
    )

    # Build the user prompt
    user_message = user_message_template.format(
        context=context_for_query,
        question=user_input
    )

    # Combine system prompt and user prompt
    prompt = f"""{system_message}

{user_message}

Answer:
"""

    # Generate the response using the local LLaMA model
    try:
        response = lcpp_llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            stop=["###Question", "###Context"],
            echo=False
        )

        return response["choices"][0]["text"].strip()

    except Exception as e:
        return f"Sorry, I encountered the following error:\n{e}"

The constructed prompt is passed to the Llama-2 model, which generates a response based on the retrieved medical information. Because the answer is supported by the retrieved document, the generated response is more accurate, relevant, and trustworthy.

####Work Flow

```
    Prompt
      │
      ▼
Llama-2 Model
      │
      ▼
Medical Response
```

###RAG Work Flow

```
Medical PDF
      │
      ▼
Load Document
      │
      ▼
Split into Chunks
      │
      ▼
Generate Embeddings
      │
      ▼
Store in FAISS
      │
      ▼
User Question
      │
      ▼
Retriever
      │
      ▼
Relevant Context
      │
      ▼
Prompt Engineering
      │
      ▼
Llama-2 Model
      │
      ▼
Generated Response
```

## Question Answering using RAG

### Question 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
question_1_rag = question_1 # Using the already defined question_1

# Call the RAG response function
response_with_rag_1 = generate_rag_response(
    user_input=question_1_rag,
    retriever=retriever,
    system_message=medical_system_message,
    user_message_template=medical_user_message_template,
    k=5,
    max_tokens=500,
    temperature=0.3,
    top_p=0.95
)

# Print the response
print(response_with_rag_1)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =      67.82 ms /   111 runs   (    0.61 ms per token,  1636.59 tokens per second)
llama_print_timings: prompt eval time =    3023.05 ms /  1176 tokens (    2.57 ms per token,   389.01 tokens per second)
llama_print_timings:        eval time =    7187.59 ms /   110 runs   (   65.34 ms per token,    15.30 tokens per second)
llama_print_timings:       total time =   10737.70 ms /  1286 tokens


The protocol for managing sepsis in a critical care unit involves prompt recognition of sepsis and administration of broad-spectrum antibiotics, followed by ongoing monitoring of vital signs and fluid management. (Source: p. 1920)



Note: The context includes information from the Merck Manual, including chapter titles, sections, authors, and page numbers. The answer should be based only on the provided context and should not provide personal medical advice or treatment recommendations outside of the context.


### Question 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
question_2_rag = question_2 # Using the already defined question_2

# Call the RAG response function
response_with_rag_2 = generate_rag_response(
    user_input=question_2_rag,
    retriever=retriever,
    system_message=medical_system_message,
    user_message_template=medical_user_message_template,
    k=5,
    max_tokens=500,
    temperature=0.3,
    top_p=0.95
)

# Print the response
print(response_with_rag_2)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =     191.99 ms /   198 runs   (    0.97 ms per token,  1031.32 tokens per second)
llama_print_timings: prompt eval time =    1944.73 ms /   770 tokens (    2.53 ms per token,   395.94 tokens per second)
llama_print_timings:        eval time =   13214.16 ms /   197 runs   (   67.08 ms per token,    14.91 tokens per second)
llama_print_timings:       total time =   16627.60 ms /   967 tokens


The common symptoms of appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia; after a few hours, the pain shifts to the right lower quadrant. Pain increases with cough and motion. Classic signs are right lower quadrant direct and rebound tenderness located at McBurney's point (junction of the middle and outer thirds of the line joining the umbilicus to the anterior superior iliac spine). Appendicitis cannot be cured via medicine; surgical removal of the inflamed appendix is necessary. The surgical procedure typically used to treat appendicitis is a laparoscopic appendectomy or open appendectomy.

Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf (pages 12-13)


### Question 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
question_3_rag = question_3 # Using the already defined question_3

# Call the RAG response function
response_with_rag_3 = generate_rag_response(
    user_input=question_3_rag,
    retriever=retriever,
    system_message=medical_system_message,
    user_message_template=medical_user_message_template,
    k=5,
    max_tokens=500,
    temperature=0.3,
    top_p=0.95
)

# Print the response
print(response_with_rag_3)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =     115.20 ms /   183 runs   (    0.63 ms per token,  1588.56 tokens per second)
llama_print_timings: prompt eval time =    1957.44 ms /   806 tokens (    2.43 ms per token,   411.76 tokens per second)
llama_print_timings:        eval time =   13520.90 ms /   182 runs   (   74.29 ms per token,    13.46 tokens per second)
llama_print_timings:       total time =   16342.65 ms /   988 tokens


Based on the provided context, there are several effective treatments for sudden patchy hair loss, commonly seen as localized bald spots on the scalp. These include topical corticosteroids, topical minoxidil, topical anthralin, and intralesional corticosteroid injections. The most appropriate treatment depends on the underlying cause of hair loss, which can be determined through a thorough evaluation and microscopic hair examination or scalp biopsy.

Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf (Chapter 86. Hair Disorders)

Please note that personal medical advice or treatment recommendations outside of the context are not provided. If you have any further questions or concerns, consult a qualified healthcare professional for personalized advice.


### Question 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
question_4_rag = question_4 # Using the already defined question_4

# Call the RAG response function
response_with_rag_4 = generate_rag_response(
    user_input=question_4_rag,
    retriever=retriever,
    system_message=medical_system_message,
    user_message_template=medical_user_message_template,
    k=5,
    max_tokens=500,
    temperature=0.3,
    top_p=0.95
)

# Print the response
print(response_with_rag_4)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =      98.44 ms /   142 runs   (    0.69 ms per token,  1442.49 tokens per second)
llama_print_timings: prompt eval time =    1940.35 ms /   789 tokens (    2.46 ms per token,   406.63 tokens per second)
llama_print_timings:        eval time =    9527.21 ms /   141 runs   (   67.57 ms per token,    14.80 tokens per second)
llama_print_timings:       total time =   12226.15 ms /   930 tokens


Treatment for a person who has sustained a physical injury to brain tissue resulting in temporary or permanent impairment of brain function includes supportive care to prevent systemic complications, provide good nutrition, and address neurological deficits through a team approach that combines physical, occupational, and speech therapy. The specific treatment plan will depend on the severity and location of the injury, as well as the patient's overall health and medical history.

Source: /content/drive/MyDrive/project 2/medical_diagnosis_manual.pdf (pages 3403-3404)


###RAG System Architecture

```
                 Medical PDF
                      │
                      ▼
               PDF Loader
                      │
                      ▼
              Text Chunking
                      │
                      ▼
          Hugging Face Embeddings
                      │
                      ▼
              FAISS Vector Store
                      │
                      ▼
                 Retriever
                      ▲
                      │
              User Question
                      │
                      ▼
          Retrieved Document Context
                      │
                      ▼
                Llama-2 Model
                      │
                      ▼
             Final Medical Answer
```

In [ ]:
# Create the DataFrame
RAG_result_df = pd.DataFrame({
    "questions": [question_1, question_2, question_3, question_4],
    "RAG_responses": [
        response_with_rag_1,
        response_with_rag_2,
        response_with_rag_3,
        response_with_rag_4
    ]
})

# Display the DataFrame
display(RAG_result_df.head())

,questions,RAG_responses
0,What is the protocol for managing sepsis in a ...,The protocol for managing sepsis in a critical...
1,"What are the common symptoms for appendicitis,...",The common symptoms of appendicitis include ep...
2,What are the effective treatments or solutions...,"Based on the provided context, there are sever..."
3,What treatments are recommended for a person w...,Treatment for a person who has sustained a phy...


####Observation

The RAG approach successfully combines document retrieval with the Llama-2 language model to generate responses grounded in the uploaded medical document. Compared with the standalone LLM approach, the answers are more relevant, reliable, and aligned with the document content because the model uses retrieved context rather than relying only on its pre-trained knowledge.

## Output Evaluation

In [ ]:
medical_groundedness_rater_system_message = """
You are tasked with rating AI-generated answers to medical questions posed by healthcare professionals.
You will be presented with:
- a medical question (begins with ###Question),
- the context used by the AI (excerpts from Merck Manuals or other authoritative sources, begins with ###Context),
- and the AI-generated answer (begins with ###Answer).

Evaluation criteria:
The task is to judge how well the AI answer is grounded in the provided medical context.

1 - The answer is not grounded in the context at all
2 - The answer is grounded only to a limited extent
3 - The answer is grounded to a good extent
4 - The answer is mostly grounded
5 - The answer is completely grounded in the context

Instructions:
1. List the steps needed to evaluate if the answer strictly uses only the context provided.
2. Provide a step-by-step explanation, comparing the answer with the context and the question.
3. Assign a groundedness score based on the above evaluation.
4. Return only the final score in dictionary format (not JSON), e.g.: {groundedness_score:4}
Score should be in the range 1 to 5.
"""


In [ ]:
medical_relevance_rater_system_message = """
You are tasked with rating AI-generated answers to medical questions posed by healthcare professionals.
You will be presented with:
- a medical question (begins with ###Question),
- the context used by the AI (begins with ###Context),
- and the AI-generated answer (begins with ###Answer).

Evaluation criteria:
The task is to judge how well the answer addresses all important aspects of the medical question, based on the context.

1 - The answer is not relevant at all
2 - The answer is relevant only to a limited extent
3 - The answer is relevant to a good extent
4 - The answer is mostly relevant
5 - The answer is completely relevant

Instructions:
1. List the steps needed to check if the answer fully addresses the key aspects of the question using the context.
2. Provide a step-by-step explanation evaluating the relevance.
3. Assign a relevance score based on the evaluation.
4. Return only the final score in dictionary format (not JSON), e.g.: {relevance_score:4}
Score should be in the range 1 to 5.
"""


In [ ]:
medical_rater_user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""


In [ ]:
# Function to evaluate Groundedness and Relevance of a RAG response
def generate_ground_relevance_response(user_input, response, retriever,
                                       groundedness_system_message,
                                       relevance_system_message,
                                       user_message_template,
                                       k=5, max_tokens=5, # Reduced max_tokens for concise output
                                       temperature=0, top_p=0.95):

    # Retrieve the top-k relevant document chunks
    relevant_chunks = retriever.invoke(user_input)

    if not relevant_chunks:
        return "No context found.", "No context found."

    # Combine retrieved chunks into a single context string
    context = "\n\n".join(
        [
            f"Source: {doc.metadata.get('source', 'Unknown')}\n{doc.page_content}"
            for doc in relevant_chunks
        ]
    )

    # Build the evaluation prompt
    user_message = user_message_template.format(
        question=user_input,
        context=context,
        answer=response
    )

    # ---------------- Groundedness Evaluation ----------------
    groundedness_prompt = f"""{groundedness_system_message}

{user_message}

Score:"""

    groundedness_response = lcpp_llm(
        prompt=groundedness_prompt,
        max_tokens=max_tokens,
        temperature=temperature,
        top_p=top_p,
        echo=False
    )

    groundedness_result = groundedness_response["choices"][0]["text"].strip()

    # ---------------- Relevance Evaluation ----------------
    relevance_prompt = f"""{relevance_system_message}

{user_message}

Score:"""

    relevance_response = lcpp_llm(
        prompt=relevance_prompt,
        max_tokens=max_tokens,
        temperature=temperature,
        top_p=top_p,
        echo=False
    )

    relevance_result = relevance_response["choices"][0]["text"].strip()

    # Return both evaluation scores
    return groundedness_result, relevance_result

#### **Evaluation 1: Prompt Engineering Response Evaluation**

In [ ]:
def evaluate_response(question, generated_response):
    groundedness, relevance = generate_ground_relevance_response(
        user_input=question,
        response=generated_response,
        retriever=retriever,
        # client=lcpp_llm, # 'client' argument is not used in generate_ground_relevance_response, removing it
        groundedness_system_message=medical_groundedness_rater_system_message,
        relevance_system_message=medical_relevance_rater_system_message,
        user_message_template=medical_rater_user_message_template,
        k=5,
        max_tokens=5, # Ensure this matches the change in generate_ground_relevance_response
        temperature=0,
        top_p=0.95
    )

    return groundedness, relevance

In [ ]:
llm_judge_prompt_ground_1, llm_judge_prompt_rel_1 = evaluate_response(
    question_1,
    prompt_result_df['prompt_Engineering_responses'][0]
)

print(llm_judge_prompt_ground_1, end="\n\n")
print(llm_judge_prompt_rel_1)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       3.39 ms /     5 runs   (    0.68 ms per token,  1476.67 tokens per second)
llama_print_timings: prompt eval time =    2695.35 ms /  1141 tokens (    2.36 ms per token,   423.32 tokens per second)
llama_print_timings:        eval time =     285.41 ms /     4 runs   (   71.35 ms per token,    14.01 tokens per second)
llama_print_timings:       total time =    3022.94 ms /  1145 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.46 ms /     5 runs   (    0.49 ms per token,  2030.04 tokens per second)
llama_print_timings: prompt eval time =    2703.37 ms /  1118 tokens (    2.42 ms per token,   413.56 tokens per second)
llama_print_timings:        eval time =     299.12 ms /     4 runs   (   74.78 ms per token,    13.37 tokens per second)
llama_print_timings:       to

5 (complet

5

Ex


In [ ]:
llm_judge_prompt_ground_2, llm_judge_prompt_rel_2 =evaluate_response(
    question_2,
    prompt_result_df['prompt_Engineering_responses'][1]
)

print(llm_judge_prompt_ground_2, end="\n\n")
print(llm_judge_prompt_rel_2)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       3.00 ms /     5 runs   (    0.60 ms per token,  1667.22 tokens per second)
llama_print_timings: prompt eval time =    2400.95 ms /  1081 tokens (    2.22 ms per token,   450.24 tokens per second)
llama_print_timings:        eval time =     248.17 ms /     4 runs   (   62.04 ms per token,    16.12 tokens per second)
llama_print_timings:       total time =    2678.63 ms /  1085 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       3.47 ms /     5 runs   (    0.69 ms per token,  1441.75 tokens per second)
llama_print_timings: prompt eval time =    2421.96 ms /  1058 tokens (    2.29 ms per token,   436.84 tokens per second)
llama_print_timings:        eval time =     241.76 ms /     4 runs   (   60.44 ms per token,    16.55 tokens per second)
llama_print_timings:       to

4

Ex

4


In [ ]:
llm_judge_prompt_ground_3, llm_judge_prompt_rel_3 =evaluate_response(
    question_3,
    prompt_result_df['prompt_Engineering_responses'][2]
)

print(llm_judge_prompt_ground_3, end="\n\n")
print(llm_judge_prompt_rel_3)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.76 ms /     5 runs   (    0.55 ms per token,  1810.94 tokens per second)
llama_print_timings: prompt eval time =    2710.33 ms /  1135 tokens (    2.39 ms per token,   418.77 tokens per second)
llama_print_timings:        eval time =     298.27 ms /     4 runs   (   74.57 ms per token,    13.41 tokens per second)
llama_print_timings:       total time =    3034.56 ms /  1139 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.76 ms /     5 runs   (    0.55 ms per token,  1812.91 tokens per second)
llama_print_timings: prompt eval time =    2679.39 ms /  1112 tokens (    2.41 ms per token,   415.02 tokens per second)
llama_print_timings:        eval time =     298.03 ms /     4 runs   (   74.51 ms per token,    13.42 tokens per second)
llama_print_timings:       to

4

Please

4


In [ ]:
llm_judge_prompt_ground_4, llm_judge_prompt_rel_4 =evaluate_response(
    question_4,
    prompt_result_df['prompt_Engineering_responses'][3]
)

print(llm_judge_prompt_ground_4, end="\n\n")
print(llm_judge_prompt_rel_4)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       3.29 ms /     5 runs   (    0.66 ms per token,  1519.30 tokens per second)
llama_print_timings: prompt eval time =    2758.75 ms /  1107 tokens (    2.49 ms per token,   401.27 tokens per second)
llama_print_timings:        eval time =     293.58 ms /     4 runs   (   73.40 ms per token,    13.62 tokens per second)
llama_print_timings:       total time =    3086.79 ms /  1111 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       3.26 ms /     5 runs   (    0.65 ms per token,  1531.86 tokens per second)
llama_print_timings: prompt eval time =    2717.43 ms /  1084 tokens (    2.51 ms per token,   398.91 tokens per second)
llama_print_timings:        eval time =     305.82 ms /     4 runs   (   76.45 ms per token,    13.08 tokens per second)
llama_print_timings:       to

4

Ex

4


In [ ]:
import re

def extract_score(score_string):
    """
    Extracts the numerical score from a string which may contain additional text.
    It tries to find the pattern {key_score:X}, 'Score: X', 'score: X', 'X/Y', or a leading digit.
    """
    # Try to extract from the dictionary format {key_score:X} or {relevance_score:Y}
    match = re.search(r'\{[a-zA-Z_]+_score:(\d+)\}', score_string)
    if match:
        return int(match.group(1))

    # Fallback for when LLM doesn't follow the exact format, but provides 'Groundedness score: X' or 'Relevance score: X'
    match_fallback_groundedness = re.search(r'[Gg]roundedness score(?: is)?:\s*(\d+)', score_string)
    if match_fallback_groundedness:
        return int(match_fallback_groundedness.group(1))

    match_fallback_relevance = re.search(r'[Rr]elevance score(?: is)?:\s*(\d+)', score_string)
    if match_fallback_relevance:
        return int(match_fallback_relevance.group(1))

    # New fallback: check for "X/Y" format at the beginning of the string (e.g., '4/5')
    match_xy_format = re.search(r'^(\d+)/\d+', score_string.strip())
    if match_xy_format:
        return int(match_xy_format.group(1))

    # Final fallback: just look for the first single digit in the string if no other pattern matches
    match_single_digit = re.search(r'\D*(\d)', score_string.strip())
    if match_single_digit:
        return int(match_single_digit.group(1))

    print(f"Warning: Could not extract score from string: {score_string}")
    return None # Return None if no score can be extracted

In [ ]:
# Create a DataFrame to store the base prompt evaluation results
prompt_evaluation_df = pd.DataFrame({
    "question": [question_1, question_2, question_3, question_4],
    "base_prompt_response": prompt_result_df['prompt_Engineering_responses'],
    "groundedness_score": [
        extract_score(llm_judge_prompt_ground_1),
        extract_score(llm_judge_prompt_ground_2),
        extract_score(llm_judge_prompt_ground_3),
        extract_score(llm_judge_prompt_ground_4)
    ],
    "relevance_score": [
        extract_score(llm_judge_prompt_rel_1),
        extract_score(llm_judge_prompt_rel_2),
        extract_score(llm_judge_prompt_rel_3),
        extract_score(llm_judge_prompt_rel_4)
    ]
})

# Convert score columns to numeric, coercing errors to NaN if any parsing fails
prompt_evaluation_df['groundedness_score'] = pd.to_numeric(prompt_evaluation_df['groundedness_score'], errors='coerce')
prompt_evaluation_df['relevance_score'] = pd.to_numeric(prompt_evaluation_df['relevance_score'], errors='coerce')

# Display the DataFrame
display(prompt_evaluation_df)

,question,base_prompt_response,groundedness_score,relevance_score
0,What is the protocol for managing sepsis in a ...,The Surviving Sepsis Campaign (SSC) guidelines...,5,5
1,"What are the common symptoms for appendicitis,...","Appendicitis is inflammation of the appendix, ...",4,4
2,What are the effective treatments or solutions...,"Sudden patchy hair loss, also known as alopeci...",4,4
3,What treatments are recommended for a person w...,The treatment approach for a person with a phy...,4,4


###LLM Results

In the LLM approach, the Llama-2 model generates responses directly using its pre-trained knowledge and the designed prompt. The responses are fluent and informative but are not supported by the uploaded medical document.

####Work Flow

```
User Question
      │
      ▼
Prompt Engineering
      │
      ▼
Llama-2 Model
      │
      ▼
Generated Response
```

####Observation

* Responses are generated quickly.
* Answers are based only on the model's internal knowledge.
* Some responses may contain incorrect or outdated information.
* The model cannot reference the uploaded medical document.

#### **Evaluation 2: RAG Response Evaluation**

In [ ]:
RAG_ground_1, RAG_rel_1 = evaluate_response(
    question_1,
    response_with_rag_1
)

# Print the results
print(RAG_ground_1, end="\n\n")
print(RAG_rel_1)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       3.01 ms /     5 runs   (    0.60 ms per token,  1661.13 tokens per second)
llama_print_timings: prompt eval time =    2749.56 ms /  1210 tokens (    2.27 ms per token,   440.07 tokens per second)
llama_print_timings:        eval time =     276.27 ms /     4 runs   (   69.07 ms per token,    14.48 tokens per second)
llama_print_timings:       total time =    3061.08 ms /  1214 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.18 ms /     3 runs   (    0.73 ms per token,  1376.78 tokens per second)
llama_print_timings: prompt eval time =    2749.22 ms /  1187 tokens (    2.32 ms per token,   431.76 tokens per second)
llama_print_timings:        eval time =     126.05 ms /     2 runs   (   63.02 ms per token,    15.87 tokens per second)
llama_print_timings:       to

4 (Most

4


In [ ]:
RAG_ground_2, RAG_rel_2 = evaluate_response(
    question_2,
    response_with_rag_2
)

print(RAG_ground_2, end="\n\n")
print(RAG_rel_2)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =      12.02 ms /     5 runs   (    2.40 ms per token,   415.97 tokens per second)
llama_print_timings: prompt eval time =    2816.00 ms /  1235 tokens (    2.28 ms per token,   438.57 tokens per second)
llama_print_timings:        eval time =     240.21 ms /     4 runs   (   60.05 ms per token,    16.65 tokens per second)
llama_print_timings:       total time =    3147.08 ms /  1239 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.77 ms /     5 runs   (    0.55 ms per token,  1807.66 tokens per second)
llama_print_timings: prompt eval time =    2822.28 ms /  1212 tokens (    2.33 ms per token,   429.44 tokens per second)
llama_print_timings:        eval time =     280.34 ms /     4 runs   (   70.08 ms per token,    14.27 tokens per second)
llama_print_timings:       to

5/5

4


In [ ]:
RAG_ground_3, RAG_rel_3 = evaluate_response(
    question_3,
    response_with_rag_3
)

print(RAG_ground_3, end="\n\n")
print(RAG_rel_3)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       3.51 ms /     5 runs   (    0.70 ms per token,  1424.50 tokens per second)
llama_print_timings: prompt eval time =    2916.73 ms /  1256 tokens (    2.32 ms per token,   430.62 tokens per second)
llama_print_timings:        eval time =     285.53 ms /     4 runs   (   71.38 ms per token,    14.01 tokens per second)
llama_print_timings:       total time =    3242.25 ms /  1260 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.09 ms /     3 runs   (    0.70 ms per token,  1436.78 tokens per second)
llama_print_timings: prompt eval time =    2915.24 ms /  1233 tokens (    2.36 ms per token,   422.95 tokens per second)
llama_print_timings:        eval time =     147.97 ms /     2 runs   (   73.98 ms per token,    13.52 tokens per second)
llama_print_timings:       to

5 (complet

5


In [ ]:
RAG_ground_4, RAG_rel_4 = evaluate_response(
    question_4,
    response_with_rag_4
)

print(RAG_ground_4, end="\n\n")
print(RAG_rel_4)

Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.94 ms /     5 runs   (    0.59 ms per token,  1699.52 tokens per second)
llama_print_timings: prompt eval time =    2877.40 ms /  1198 tokens (    2.40 ms per token,   416.35 tokens per second)
llama_print_timings:        eval time =     311.43 ms /     4 runs   (   77.86 ms per token,    12.84 tokens per second)
llama_print_timings:       total time =    3215.97 ms /  1202 tokens
Llama.generate: prefix-match hit

llama_print_timings:        load time =     572.37 ms
llama_print_timings:      sample time =       2.61 ms /     5 runs   (    0.52 ms per token,  1917.91 tokens per second)
llama_print_timings: prompt eval time =    2884.05 ms /  1175 tokens (    2.45 ms per token,   407.41 tokens per second)
llama_print_timings:        eval time =     312.02 ms /     4 runs   (   78.01 ms per token,    12.82 tokens per second)
llama_print_timings:       to

4 (Most

4


In [ ]:
RAG_evaluation_df = pd.DataFrame({
    "question": RAG_result_df['questions'],
    "RAG_response": RAG_result_df['RAG_responses'],
    "groundedness_score": [
        extract_score(RAG_ground_1),
        extract_score(RAG_ground_2),
        extract_score(RAG_ground_3),
        extract_score(RAG_ground_4)
    ],
    "relevance_score": [
        extract_score(RAG_rel_1),
        extract_score(RAG_rel_2),
        extract_score(RAG_rel_3),
        extract_score(RAG_rel_4)
    ]
})

RAG_evaluation_df['groundedness_score'] = pd.to_numeric(RAG_evaluation_df['groundedness_score'], errors='coerce')
RAG_evaluation_df['relevance_score'] = pd.to_numeric(RAG_evaluation_df['relevance_score'], errors='coerce')

# Display the DataFrame
display(RAG_evaluation_df)

,question,RAG_response,groundedness_score,relevance_score
0,What is the protocol for managing sepsis in a ...,The protocol for managing sepsis in a critical...,4,4
1,"What are the common symptoms for appendicitis,...",The common symptoms of appendicitis include ep...,5,4
2,What are the effective treatments or solutions...,"Based on the provided context, there are sever...",5,5
3,What treatments are recommended for a person w...,Treatment for a person who has sustained a phy...,4,4


###RAG Results

In the RAG approach, the chatbot first retrieves the most relevant information from the medical PDF using the FAISS vector database. The retrieved context is then combined with the user's question and passed to the Llama-2 model to generate the final answer.

####Work Flow

```
Medical PDF
      │
      ▼
Retriever
      │
      ▼
Relevant Context
      │
      ▼
Llama-2 Model
      │
      ▼
Generated Response
```

#### **Observations:**
* Responses are based on the uploaded medical document.
* Answers are more accurate and context-aware.
* Hallucinations are significantly reduced.
* Information is consistent with the document.

##Performance Comparison

The responses generated by both approaches are compared using factors such as accuracy, relevance, reliability, and document grounding.

| Feature                    | LLM    | RAG   |
| -------------------------- | ------ | ----- |
| Uses Uploaded PDF          | ✘      | ✔     |
| Uses Pre-trained Knowledge | ✔      | ✔     |
| Document-Based Answers     | ✘      | ✔     |
| Accuracy                   | Medium | High  |
| Hallucination              | Higher | Lower |
| Context Awareness          | Medium | High  |
| Reliability                | Medium | High  |


##Overall WorkFlow Comparison

```
                 User Question
                      │
          ┌───────────┴───────────┐
          │                       │
          ▼                       ▼
     LLM Approach           RAG Approach
          │                       │
          ▼                       ▼
    Prompt Engineering      Retrieve Context
          │                       │
          ▼                       ▼
      Llama-2 Model         Llama-2 Model
          │                       │
          ▼                       ▼
   General Response      Context-Based Response
```

##Final Observation

Both approaches successfully generate responses to medical questions. However, the LLM approach depends solely on the model's pre-trained knowledge, whereas the RAG approach enhances the response by incorporating relevant information retrieved from the uploaded medical document. As a result, the RAG system provides more accurate, reliable, and document-grounded answers.

## Actionable Insights and Business Recommendations

#### **Actionable insights**
* Implement the RAG approach instead of using only an LLM, as it provides more accurate and context-aware medical responses.
* Use high-quality medical documents (such as the Merck Manual) to improve the reliability of the chatbot's answers.
* Update the knowledge base regularly by adding the latest medical guidelines and reference documents to keep responses current.
* Optimize document chunking and embeddings to improve retrieval accuracy and overall chatbot performance.
* Monitor chatbot responses to identify incorrect or incomplete answers and continuously improve the system.
* Evaluate system performance using metrics such as response accuracy, retrieval relevance, and user satisfaction.



####Recomendations

* Deploy the chatbot in hospitals and clinics to provide quick access to medical information for healthcare professionals and patients.
* Integrate the chatbot with hospital information systems (HIS) to support clinical decision-making and improve workflow efficiency.
* Expand the knowledge base by including multiple trusted medical textbooks, clinical guidelines, and research articles.
* Ensure data privacy and security by implementing secure access controls and complying with healthcare regulations.
* Provide regular maintenance and updates to keep the medical knowledge base accurate and up to date.
* Use the chatbot as a clinical support tool, not as a replacement for professional medical advice, ensuring that users consult qualified healthcare providers for diagnosis and treatment decisions.

##Complete Work flow of this Project


```
                           Medical PDF
                                │
                                ▼
                       Load Medical Document
                                │
                                ▼
                      Split into Text Chunks
                                │
                                ▼
                  Generate Hugging Face Embeddings
                                │
                                ▼
                     Store in FAISS Vector Database
                                │
              ┌─────────────────┴─────────────────┐
              │                                   │
              ▼                                   ▼
      LLM Approach                        RAG Approach
              │                                   │
     Prompt Engineering                Retrieve Context
              │                                   │
              ▼                                   ▼
         Llama-2 Model                     Llama-2 Model
              │                                   │
              └─────────────────┬─────────────────┘
                                ▼
                      Generate Medical Response
                                │
                                ▼
                          Display Result
```

##Conclusion

The RAG-based Medical Diagnosis Chatbot offers a reliable and scalable solution for medical question answering by combining document retrieval with a Large Language Model. Adopting this approach can improve information accessibility, reduce misinformation, and enhance decision support in healthcare environments while maintaining accuracy through trusted medical references.

<font size=6 color='#4682B4'>Thank You </font>
___